### PR TITLE
Make `tctl <resource> ls` command outputs consistent

### PR DESCRIPTION
**Pull request by lxea**
This makes tctl commands consistent between  different outputs as described in #9379. This adds a kube ls command.

### CODE REVIEW

**Review by lxea — COMMENTED** [review]
In order to get Teleport version and host,  GetWindowsDesktopServices must be used however I'm not sure how to associate a windows desktop with a windows desktop service. Any ideas @russjones ?

**Comment by russjones** [comment]
@lxea @zmb3 should be able to help you out on the Windows Desktop questions.

**Comment by zmb3** [comment]
@lxea there currently isn't an association between a Windows Desktop and the Windows Desktop Service (or services) that can connect to it.

We'll be adding this soon (feel free to follow #9277 for progress).

If there's anything in particular you need in the implementation to make this work easier let me know, but the plan is basically do make it work exactly like database access.

File: tool/tctl/common/collection.go
Comment (zmb3) [inline tool/tctl/common/collection.go:726]: I'm not sure I would say "Public Address" here. "Public" may be seen as routable and that's often not going to be true.

File: tool/tctl/common/collection.go
Comment (zmb3) [inline tool/tctl/common/collection.go:728]: Maybe consider hiding any labels with a `teleport.dev/` prefix? Those are the ones that are auto-populated via LDAP and likely not important to something browsing via `tctl`.

File: tool/tctl/common/desktop_command.go
Comment (zmb3) [inline tool/tctl/common/desktop_command.go:46]: In most cases, we're referring to these resources as "Windows Desktops" as desktop access was designed with the idea that we may support other OS in the future.

That said, `tctl windows_desktop ls` is kind of a pain to type, so maybe we just make it `desktops ls` and add an `OS` column in the future if we support other desktops.

**Comment by russjones** [comment]
@lxea Can we use a similar approach to what we did with `tsh` where we size the output to the size of the terminal? I think that will let us show more labels whenever we can.

File: tool/tctl/common/collection.go
Comment (lxea) [reply tool/tctl/common/collection.go:791]: I'm not sure I understand this change, I've rearranged it so instead its just deletes unnecessary labels if its in verbose mode so there isn't an if else branch and added a comment which I think looks, at least, a bit clearer than before.

Comment (Tener) [reply tool/tctl/common/collection.go:791]: Thank you, this looks cleaner indeed. I think there is a bug though, as we first assign result of `GetAllLabels()` to `labels`, but then we discard it by assigning an empty map `labels = map[string]string{}`. 

What I meant originally was something along the lines of (untested code):
```
		labels := map[string]string{}
		for key, value := range d.desktop.GetAllLabels() {
			// skip teleport.dev labels, unless we're in verbose mode.
			if verbose || !strings.HasPrefix(key, teleportNamespace) {
				labels[key] = value
			}
		}
```

Comment (lxea) [reply tool/tctl/common/collection.go:790]: Oh, I see, I think i prefer the version with deleting, I've remove the line that was clearing the labels that got left in too.

**Comment by r0mant** [comment]
@Tener @zmb3 Since you guys looked at this before, mind re-reviewing?

File: tool/tctl/common/collection.go
Comment (Tener) [inline tool/tctl/common/collection.go:38]: Should it be in shared place instead, like `api/types/constants.go`?

File: tool/tctl/common/desktop_command.go
Comment (Tener) [inline tool/tctl/common/desktop_command.go:86]: Nit: `trace.Wrap()` is missing here and in analogous other places (where we simply assign err, without checking it). I believe it is most accurate to apply here and simply leave `return err` below.

File: tool/tctl/main.go
Comment (Tener) [inline tool/tctl/main.go:35]: The same needs to be done in Enterprise implementation of tctl in `teleport.e` repo.

File: tool/tctl/common/kube_command.go
Comment (Tener) [inline tool/tctl/common/kube_command.go:51]: Can we add JSON to the mixture of formats? Some other commands support it, and I believe it is easier to process programmatically by CLI tools.

File: tool/tctl/common/desktop_command.go
Comment (Tener) [inline tool/tctl/common/desktop_command.go:52]: As with kube command: I believe JSON support is worthwhile.

File: tool/tctl/common/desktop_command.go
Comment (Tener) [inline tool/tctl/common/desktop_command.go:89]: At first sight, it feels strange we are issuing `.writeYaml()` against a different collection than the one with text format. This is different from other commands and suggests an opportunity for simplification.

**Comment by Tener** [comment]
@r0mant @lxea sorry for the delay. I have reviewed everything now, including the code I didn't look at close enough on the first pass. Sorry for pointing out stuff I could have found already in the past but missed...

**Review by Tener — APPROVED** [review]
LGTM

**Comment by Tener** [comment]
@r0mant @zmb3 I've approved, but scripts require your approval too.

File: api/types/app.go
Comment (r0mant) [inline api/types/app.go:246]: Hmm, I think `a.Version` represents the _resource_ version, not Teleport version. There is also `GetVersion()` method that returns the same.

Applications themselves don't have information about Teleport version. AppServer resources (which represent an application proxied by a particular Teleport agent) do.

So I don't think this method is neede? tctl apps ls already display AppServers IIRC.

File: tool/tctl/common/collection.go
Comment (r0mant) [inline tool/tctl/common/collection.go:54]: Instead of threading an additional boolean flag into each writeXX method for each resource, can you make `verbose bool` a field of a particular "Collection" object? This way it will only be used by collections that need/support it and we won't have to update all of them to pass additional parameter which is most of the time unused.

**Review by r0mant — CHANGES_REQUESTED** [review]
Have a couple suggestions.

File: api/types/app.go
Comment (lxea) [reply api/types/app.go:246]: Yeah this wasnt needed, just a holdover from when I was trying some stuff, I'm using appserver anyway, removed.

File: api/types/constants.go
Comment (r0mant) [inline api/types/constants.go:302]: I think we use this label quite a bit throughout the code. If we're introducing a constant for it, I feel like we should go ahead and use it in all places that currently use `teleport.dev`. Otherwise you'd have to remember to change it everywhere. So ideally the only place "teleport.dev" string should appear verbatim is in this constant definition.

File: lib/asciitable/table.go
Comment (r0mant) [inline lib/asciitable/table.go:66]: Nit: Instead of making a separate function, what do you think about updating MakeTable to support accepting variable length argument with rows? Current invocations won't change and we'll have one less function.

```go
func MakeTable(headers []string, rows ...[]string) Table {
  // existing impl.
  for _, row := range rows {
    t.AddRow(row)
  }
  return t
}
```

File: lib/asciitable/table.go
Comment (r0mant) [inline lib/asciitable/table.go:74]: Add godoc.

File: tool/tctl/common/collection.go
Comment (r0mant) [inline tool/tctl/common/collection.go:142]: I'm not 100% in which case we set the server's public address, but I think it only may make sense for proxies. I think this collection represents SSH nodes so we probably only want to show Addr here.

File: tool/tctl/common/collection.go
Comment (r0mant) [inline tool/tctl/common/collection.go:804]: I would make this into a helper method and used for displaying other resources as well since they also can have teleport.dev labels, so it's consistent.

File: tool/tctl/common/node_command.go
Comment (r0mant) [inline tool/tctl/common/node_command.go:248]: I'm pretty sure we have constants for `"text"` and `"yaml"`.

File: tool/tctl/common/node_command.go
Comment (r0mant) [inline tool/tctl/common/node_command.go:89]: Use the existing constant for `"text"`.

File: tool/tctl/common/node_command.go
Comment (r0mant) [inline tool/tctl/common/node_command.go:257]: What about json?

**Review by r0mant — APPROVED** [review]
Approving but please address suggestions I left before merging.

**Comment by TwiN** [comment]
While `MakeTableWithTruncatedColumn` does make the outputs consistent, it also broke automations that extracted the name from ls commands 😔

**Comment by Tener** [comment]
@TwiN 

> While `MakeTableWithTruncatedColumn` does make the outputs consistent, it also broke automations that extracted the name from ls commands 😔

`--format=json` switch is there for automation purposes. For example, if you want to extract database names, you can do:

`tctl db ls --format=json | jq '.[].metadata.name'`

A great number of other commands support `--format=json`, and the output format rarely changes. If there is a command which doesn't support that flag feel free to raise a [feature request](https://github.com/gravitational/teleport/issues/new?assignees=&labels=feature-request&template=feature_request.md) and tell us about your use case.

**Comment by TwiN** [comment]
@Tener Good to know! I must've missed the `--format` flag!
Thanks :)

**Pull request merged**
